### DISCUSSION TITLE
Extend the functionalities of Lawn_mower

### DISCUSSION THREAD

**Discussion by cociweb** [home-assistant/architecture#1121] [Entity Models] [answered]
Hello 
According to the request in https://github.com/home-assistant/core/pull/123093#issuecomment-2266601870 by @MartinHjelmare  please approve the architectural changes of lawn_mower integration.

Introduced new -optional- functionalities/services are:
- stop then cancel mowing task
- continue/resume an ongoing task
- fixed/spot mowing

Currently available tasks are only start_mowing, pause, and dock.

**Comment by MartinHjelmare** [comment]
Please add more context. What do the services do, why are the new services required and what integrations would be able to leverage the new services?

Eg what's the difference between stop and dock? Why can't start be used to continue? That's how we have documented these services already.

https://developers.home-assistant.io/docs/core/entity/lawn-mower#methods

**Comment by cociweb** [comment]
Allow me to outline a draft context for Tuya lawnmower devices, referencing my information from the developer portal.
Actions (formarly services):
![Képernyőkép ekkor: 2024-08-03 21-25-04](https://github.com/user-attachments/assets/719c7598-c291-47ef-b048-0e3366b2cde0)
And status values:
![image](https://github.com/user-attachments/assets/f0b2449c-5403-442d-9541-ef0475ab687d)
Please be aware that [Errors](https://github.com/user-attachments/assets/bbc4e17c-0313-4166-8076-45eaf7a6460d) and [Warnings](https://github.com/user-attachments/assets/7b6675e9-a1df-40a0-93bf-88b9c0eb78af) are managed individually and are not within the scope of this PR/arch.

Or you can find more reversed and documented stuff on  [pymoebot](https://github.com/Whytey/pymoebot?tab=readme-ov-file#tuya-data-points) or on  [tinytuya](https://github.com/Whytey/pymoebot?tab=readme-ov-file#tuya-data-points)

I believe they are not specific to any device and represent a very simple subset without any unnecessary elements.

Lawnmowers essentially serve as an equivalent to [sweepers/vacuum](https://github.com/home-assistant/architecture/issues/29)  from this perspective, with the primary distinction being zone management. This differentiation has been noted in the [initial architecture document](https://github.com/home-assistant/architecture/discussions/894) of the lawn mower. Zone handling is not within the scope at this time. Careful reading will reveal that start is not equal to resume, pause is not equal to cancel, and starting mowing is not the same as fixed mowing; this is the narrative behind it.

> What do the services do

They proceed with the situation as it is. In the following PR you can find the requested information: https://github.com/home-assistant/home-assistant.io/pull/34100 (which was already available on the same time as the code PR.)

Please be aware that the original concept has not been rewritten or overwritten; it has merely been expanded to accommodate additional requirements.

That's all I can add and make. I'm very pleased to have my PR prepared as it stands. The rest is up to you!

**Reply by MartinHjelmare** [reply]
> Careful reading will reveal that start is not equal to resume, pause is not equal to cancel, and starting mowing is not the same as fixed mowing

What do you mean with this? You'll need to explain in this thread why start can't be used instead of resume when the device was paused already. The same for why dock can't be used for stop/cancel.

One example of an integration is normally not enough to add new entity features.

**Reply by cociweb** [reply]
:)
I don't really understand what is not clear.
Start: start a new mowing task.
Pause: hangs an ongoing task
Resume: continue the previously hanged task.
Cancel: stops and clear the previous task
Dock: start a new task to find the closes edge then goes to the station.
fixed mowing: start a new task where the device is spinning around with increased radius. In my case: after a certain time it automatically starts an automaticall mowing task.

In my case, you are not able to start a new task, until the previous is not deleted maybe some que is used.

That's all I can add and make.

**Reply by MartinHjelmare** [reply]
The implementing integration can handle the case that the mower continues a previous task, if start is called when there's already a task in progress.  This is already how we have documented `start_mowing`. Same relationship can be said for dock and stop/cancel.

We don't have a concept of task queue in the lawn mower integration currently. We only talk about max one task, the task that is in progress. If a task queue should be supported, a separate discussion should be had around that.

**Reply by cociweb** [reply]
> This is already how we have documented

I created the PR because the current approach lacks maturity. 

With stop and resume actions, there's no need for a different concept to manage queues - we don't need to reinvent the wheel. :) 

The PR is prepared to address the flawed concept and has been tested on my end. If it turns out you don't need it, then close it in short and it will become a custom integration.

**Reply by MartinHjelmare** [reply]
My point is that the reason for the addition of stop/cancel and resume seem to be to manage task queues. If we only have one task, and no task queue, we can assume that dock is the same as stop/cancel and start is the same as resume when there's a task already in progress.

I think we should keep the lawn mower simple and not go into managing task queues.

**Reply by mikey0000** [reply]
FWIW mammotion integration resumes when you press the start mow button if the current task is paused. Visually the mower UI shows that the mower is paused.

I've had to implement additional custom services since the default ones for start mowing are too simple and do not support data (unless I'm mistaken which I would be more than happy to be) since a mammotion mower requires pre-configuration before starting a mow.

Cancel task can be implemented as a custom service to the integration as well. That one is less useful as you can also run return to dock, which will stop the current task, but it won't cancel it.

Really need to start with one item to tackle and validate that it indeed is the same pain point across the majority of mowers to benefit everyone.

I'm going to have to implement a few services to cover functionality for mammotion, but I consider this fine since these aren't the same across all mowers.

**Comment by andyb2000** [comment]
I would welcome this, I mistakenly did a PR doing a very simple addition to "returning to dock" but I see this is much more extensive and would be useful for mower components, so I support this.

We certainly need to include the additional entries to const.py for the different entity states.

FYI the simple change I wished for is in https://github.com/andyb2000/homeassistant-core/commit/05b7571f1b27dceb1e6985b0443864a35319de6d
which avoids the additional complication of moving too close into component changes, but allows better state information similar to the entries in this PR for const.py

**Reply by joostlek** [reply]
As this PR is already answered and is more targeted towards a task queue, while your discussion is more specific. Supporting it like this makes it harder to discuss and approve your change specifically

**Reply by cociweb** [reply]
The PR is still over there. This is just a discussion, where others can reflect their additional needs to make the correct decision.  The PR can be reopen anytime, since that is ready for use.... so +1 are always welcome.

**Reply by mikey0000** [reply]
@andyb2000 your change has been implemented. returning is a thing as of 2024.9 / 10

**Comment by emontnemery** [comment]
I'm closing this because it's quite old, and it overlaps with the approved proposal in #1376

Please feel free to create new proposals for the non overlapping parts if they are still relevant 👍

**Reply by cociweb** [reply, marked as answer]
:D
